src/python/pants/backend/docker/docker_build_context.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+from dataclasses import dataclass
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.core.goals.package import BuiltPackage, PackageFieldSet
+from pants.core.target_types import FilesSources, ResourcesSources
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import AddPrefix, Digest, MergeDigests
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    Sources,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DockerBuildContext:
+    digest: Digest
+
+
+@dataclass(frozen=True)
+class DockerBuildContextRequest:
+    address: Address
+    context_root: str
+
+
+@rule
+async def create_docker_build_context(request: DockerBuildContextRequest) -> DockerBuildContext:
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest([request.address]))
+
+    # Get all sources, going into the build context.
+    sources_request = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[t.get(Sources) for t in transitive_targets.closure],
+            for_sources_types=(DockerImageSources, FilesSources, ResourcesSources),
```

Comment:
I'll do some experiments..

Reply:
On more reflection, I think it's OK to merge this as-is and revisit later. There are two issues here:

1. This may be the "wrong" way to get the Dockerfile into the snapshot that we show to the `docker` binary, but the Dockerfile does need to be in that snapshot, so it's mostly semantics whether we call that snapshot the "context" snapshot or the "context+Dockerfile" snapshot. I.e., this does at least work correctly, even if the semantics are slightly off.

2. True, this will pick up the `Dockerfile` sources of all other `docker_image` targets that this one depends on, which we don't want it to. But it will also pick up the `resources` and `files` deps of those other targets, which it also shouldn't. So we have to solve that problem some other way anyway...

Reply:
1. Agree. I've been using Docker for so long (2015), and I'm sure in the early days, is what I'm still relying on, but they have certainly improved upon that, as now it doesn't seem to have to be remotely related where the Dockerfile is located, in relation to the docker build context. It works regardless. So I take it that you've read the docs correct, and I misinterpreted the note I found (I'm assuming that the docker client handles the Dockerfile in the context transparently from the user.)
2. Will look into that, see if I can make any sense of it..

Reply:
Opened issue #12694 to track these two points.